tools/autograd/derivatives.yaml

```diff
@@ -332,6 +332,11 @@
   mat2: self.transpose(1, 2).conj().bmm(grad)
   result: self_t.bmm(mat2_p) + self_p.bmm(mat2_t)
 
+- name: matmul(Tensor self, Tensor other) -> Tensor
+  self: at::matmul(grad, other.transpose(-2, -1).conj())
+  other: at::matmul(self.transpose(-2, -1).conj(), grad)
+  result: at::matmul(self_t, other_p) + at::matmul(self_p, other_t)
```

Comment:
This looks questionable. How come we need a new derivative definition for matmul after this PR. Aren't you just splitting matmul into out/non-out versions

Reply:
Because the `non_out` version dispatches into the non-differentiable `_out` version. See also https://github.com/pytorch/pytorch/pull/66118#issuecomment-936311883